pkg/endpoint/bpf.go

```diff
@@ -228,88 +228,77 @@ func (p *proxyPolicy) GetListener() string {
 	)
 
-	changes := policy.ChangeState{
-		Adds: make(policy.Keys),
-		Old:  make(policy.MapStateMap),
-	}
-
 	// create or update proxy redirects
-	e.desiredPolicy.UpdateRedirects(ingress,
-		func(l4 *policy.L4Filter) map[string]uint16 {
-			for _, v := range l4.PerSelectorPolicies {
-				if v == nil || !v.IsRedirect() {
-					continue
-				}
-				// proxyID() returns also the destination port for the policy,
-				// which may be resolved from a named port
-				proxyID, dstPort, dstProto := e.proxyID(l4, v.Listener)
-				if proxyID == "" {
-					// Skip redirects for which a proxyID cannot be created.
-					// This may happen due to the named port mapping not
-					// existing or multiple PODs defining the same port name
-					// with different port values. The redirect will be created
-					// when the mapping is available or when the port name
-					// conflicts have been resolved in POD specs.
-					continue
-				}
-				// desiredRedirects starts out empty, so we can use it check
-				// if the redirect has already been updated on this round.
-				if desiredRedirects[proxyID] != 0 {
-					continue
-				}
-
-				var redirectPort uint16
-
-				pp := e.newProxyPolicy(l4, v, dstPort, dstProto)
-				proxyPort, err, finalizeFunc, revertFunc := e.proxy.CreateOrUpdateRedirect(e.aliveCtx, &pp, proxyID, e, proxyWaitGroup)
-				if err != nil {
-					// Skip redirects that can not be created or updated.  This
-					// can happen when a listener is missing, for example when
-					// restarting and k8s delivers the CNP before the related
-					// CEC.
-					// Policy is regenerated when listeners are added or removed
-					// to fix this condition when the listener is available.
-					e.getLogger().WithField(logfields.Listener, pp.GetListener()).WithError(err).Debug("Redirect rule with missing listener skipped, will be applied once the listener is available")
-					continue
-				}
-				redirectPort = proxyPort
-				finalizeList.Append(finalizeFunc)
-				revertStack.Push(revertFunc)
-				desiredRedirects[proxyID] = redirectPort
+	selectorPolicy.CreateRedirects(func(l4 *policy.L4Filter) error {
+		for _, v := range l4.PerSelectorPolicies {
+			if v == nil || !v.IsRedirect() {
+				continue
+			}
+			// proxyID() returns also the destination port for the policy,
+			// which may be resolved from a named port
+			proxyID, dstPort, dstProto := e.proxyID(l4, v.Listener)
+			if proxyID == "" {
+				// Skip redirects for which a proxyID cannot be created.
+				// This may happen due to the named port mapping not
+				// existing or multiple PODs defining the same port name
+				// with different port values. The redirect will be created
+				// when the mapping is available or when the port name
+				// conflicts have been resolved in POD specs.
+				continue
+			}
+			// desiredRedirects starts out empty, so we can use it check
+			// if the redirect has already been updated on this round.
+			if desiredRedirects[proxyID] != 0 {
+				continue
```

Comment:
if `e.proxy.CreateOrUpdateRedirect()` have access to both the `L4Policy` and the `PerSelectorPolicy`, why is it safe to skip here? This seems like it could cause breakage -- what if `CreateOrUpdateRedirect()` reads from more fields than is in the `proxyID`? We could have a misconfiguration.

Reply:
I suspect we need to refactor `CreateOrUpdateRedirect()` to get everything it needs from the ProxyID?

Reply:
`PerSelectorPolicy` is only used to access the listener name, which is also part of the `proxyID`. I'll change the `proxyPolicy` implementation to make this clear.

Reply:
much cleaner, thanks!